### PR TITLE
refactor: move agent_final_response to utility function with proper type checking

### LINKED ISSUE #918: refactor: agent_final_response
We should refactor this function in openhands/sdk/conversation/base.py

instead of using hasattr/getattr anywhere, we should just use isinstance(ActionEvent) to do the actual type checks.

Also i think we should not put this in the base class of Conversation, rather we should put it as an util function for conversation.

```python
    def agent_final_response(self) -> str:
        """Extract the final response from the agent.

        An agent can end a conversation in two ways:
        1. By calling the finish tool
        2. By returning a text message with no tool calls

        Returns:
            The final response message from the agent, or empty string if not found.
        """
        # Find the last finish action or message event from the agent
        for event in reversed(self.state.events):
            # Case 1: finish tool call
            if (
                type(event).__name__ == "ActionEvent"
                and hasattr(event, "source")
                and getattr(event, "source") == "agent"
                and hasattr(event, "tool_name")
                and getattr(event, "tool_name") == "finish"
            ):
                # Extract message from finish tool call
                if hasattr(event, "action") and hasattr(
                    getattr(event, "action"), "message"
                ):
                    message = getattr(getattr(event, "action"), "message")
                    return message
                else:
                    break
            # Case 2: text message with no tool calls (MessageEvent)
            elif (
                type(event).__name__ == "MessageEvent"
                and hasattr(event, "source")
                and getattr(event, "source") == "agent"
                and hasattr(event, "llm_message")
            ):
                llm_message = getattr(event, "llm_message")
                text_parts = content_to_str(llm_message.content)
                return "".join(text_parts)
        return ""

```

### PR DESCRIPTION
**Pull request by xingyaoww**
## Summary
This PR refactors the `agent_final_response` method to address the requirements in issue #918.

## Changes
- **Created `response_utils.py`**: New utility module containing the `agent_final_response` function
- **Replaced `hasattr/getattr` with `isinstance` checks**: Now using proper type checking with `isinstance(ActionEvent)` and `isinstance(MessageEvent)` instead of string-based type checking
- **Moved logic from base class to utility function**: Extracted the logic from `BaseConversation.agent_final_response()` method to a standalone utility function for better code organization
- **Added comprehensive test coverage**: Created 10 test cases covering various scenarios including:
  - Finish action extraction
  - Message event extraction
  - Mixed events with precedence
  - Edge cases (empty events, None action, non-agent sources, etc.)
- **Exported from conversation module**: Added `agent_final_response` to the conversation module's public API

## Testing
- All existing tests pass (219 tests in `tests/sdk/conversation/`)
- Added 10 new test cases specifically for `agent_final_response`
- Pre-commit checks pass for all source files (response_utils.py, base.py, __init__.py)

## Related Issue
Closes #918

## Code Quality
- Follows repository conventions for type checking and code organization
- No hasattr/getattr usage - all type checking done with isinstance
- Cleaner separation of concerns by moving utility logic out of base class

@xingyaoww can click here to [continue refining the PR](https://app.all-hands.dev/conversations/0e0b51b1790b4bf89d6bd1a232089cf9)

---
**Agent Server images for this PR**

• **GHCR package:** https://github.com/OpenHands/agent-sdk/pkgs/container/agent-server

**Variants & Base Images**
| Variant | Base Image | Docs / Tags |
|---|---|---|
| golang | `golang:1.21-bookworm` | [Link](https://hub.docker.com/_/golang:1.21-bookworm) |
| java | `eclipse-temurin:17-jdk` | [Link](https://hub.docker.com/_/eclipse-temurin:17-jdk) |
| python | `nikolaik/python-nodejs:python3.12-nodejs22` | [Link](https://hub.docker.com/_/nikolaik/python-nodejs:python3.12-nodejs22) |

**Pull (multi-arch manifest)**
```bash
docker pull ghcr.io/openhands/agent-server:2cc0d8e-python
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  --name agent-server-2cc0d8e-python \
  ghcr.io/openhands/agent-server:2cc0d8e-python
```

**All tags pushed for this build**
```
ghcr.io/openhands/agent-server:2cc0d8e-golang
ghcr.io/openhands/agent-server:v1.0.0a4_golang_tag_1.21-bookworm_binary
ghcr.io/openhands/agent-server:2cc0d8e-java
ghcr.io/openhands/agent-server:v1.0.0a4_eclipse-temurin_tag_17-jdk_binary
ghcr.io/openhands/agent-server:2cc0d8e-python
ghcr.io/openhands/agent-server:v1.0.0a4_nikolaik_s_python-nodejs_tag_python3.12-nodejs22_binary
```

_The `2cc0d8e` tag is a multi-arch manifest (amd64/arm64); your client pulls the right arch automatically._

### CODE REVIEW

File: openhands-sdk/openhands/sdk/conversation/base.py [hunk 1]
```diff
@@ -146,46 +146,6 @@ def get_persistence_dir(
         """Get the persistence directory for the conversation."""
         return str(Path(persistence_base_dir) / conversation_id.hex)
 
-    def agent_final_response(self) -> str:
```
Comment (enyst) [inline openhands-sdk/openhands/sdk/conversation/base.py:149]: Are we sure about this? I feel like we need a Conversation API that makes it easier / expected for client developers to work with, and the final agent message is fairly important, maybe important enough to offer an easy way to get it. 🤔 

Not a strong opinion, I just want to pick your brain, how do you think about it?

Comment (xingyaoww) [reply openhands-sdk/openhands/sdk/conversation/base.py:149]: I'm just not super happy about having it return a string (which does not contain the complete information), hence i'd propose to move it to a util function.

And overall i think we will have a lot of helper function like this down the road and i don't really want to clutter the Conversation base class with all these helper functions to make it hard to read 😭 

And as an SDK developer, it is probably helpful for them to understand the difference between `MessageEvent` and `AgentFinishEvent` anyway (if not, we should just merge both). What do you think?

Comment (enyst) [reply openhands-sdk/openhands/sdk/conversation/base.py:149]: Ahh, right, a public method on Conversation should probably return the event, not a string.

I hear you on helper functions, yes, no need to keep them in Conversation.
